test/e2e/engine/template.go

```diff
@@ -161,8 +162,16 @@ func Build(cfg *config.Config, masterSubnetID string, agentSubnetIDs []string, i
 		prop.ServicePrincipalProfile.ObjectID = config.ClientObjectID
 	}
 
-	if prop.OrchestratorProfile.KubernetesConfig == nil {
-		prop.OrchestratorProfile.KubernetesConfig = &vlabs.KubernetesConfig{}
+	var version string
+	if prop.OrchestratorProfile.OrchestratorRelease != "" {
+		version = prop.OrchestratorProfile.OrchestratorRelease + ".0"
+	} else if prop.OrchestratorProfile.OrchestratorVersion != "" {
+		version = prop.OrchestratorProfile.OrchestratorVersion
+	}
+	if common.IsKubernetesVersionGe(version, "1.12.0") {
+		if prop.OrchestratorProfile.KubernetesConfig == nil {
```

Comment:
why do we only want to set a previously nil `KubernetesConfig` when k8s version >= 1.12.0?

Reply:
Because only in the >= 1.12.0 scenario do we want the E2E runner to explicitly inject the below `controllerManagerConfig` values into the api model before we generate/deploy. The nil check is just to anticipate that the input api model may have a nil `kubernetesConfig`, and so we have to create it from scratch in such cases in order to define the child `controllerManagerConfig` object.

Reply:
The high level purpose here is to optimize the E2E test time without having to explicitly update all of our test api models in order to do so. The default "wait until deciding to HPA scale down" config is 5 minutes, which basically means our "validate that HPA scaled down after load stopped" test takes at least that long. This (valid) configuration just means that our test clusters are configured for a more responsive/aggressive HPA scale down behavior, which will speed up the test runs.